linera-rpc/src/cross_chain_message_queue.rs

```diff
@@ -0,0 +1,199 @@
+                    }
+                }
+
+                Action::Retry => {
+                    linera_base::time::timer::sleep(cross_chain_retry_delay * state.retries).await;
+                    Action::Proceed { id: state.id }
+                }
+            },
+        )
+    };
+
+    loop {
+        tokio::select! {
+            Some((queue, action)) = futures.next() => {
+                let Entry::Occupied(mut state) = job_states.entry(queue) else {
+                    panic!("running job without state");
+                };
+
+                let remove = matches!(action, Action::Proceed { id } if state.get().id < id)
+                    || matches!(action, Action::Retry if state.get().retries >= cross_chain_max_retries);
+                if remove {
+                    state.remove();
+                    continue;
+                }
+
+                if let Action::Retry = action {
+                    state.get_mut().retries += 1
+                }
+
+                futures.push(run_action.clone()(action, queue, state.get().clone()));
+            }
+
+            request = receiver.next() => {
+                let Some((request, shard_id)) = request else { break };
+                #[cfg(with_metrics)]
+                CROSS_CHAIN_MESSAGE_TASKS.set(job_states.len() as i64);
+
+                if rand::thread_rng().gen::<f32>() < cross_chain_sender_failure_rate {
+                    warn!("Dropped 1 cross-chain message intentionally.");
+                    continue;
+                }
+
+                let queue = QueueId::new(&request);
+
+                let task = Task {
+                    shard_id,
+                    request,
+                };
+
+                match job_states.entry(queue) {
+                    Entry::Vacant(entry) => futures.push(run_action.clone()(
+                        Action::Proceed { id: 0 },
+                        queue,
+                        entry.insert(Job {
+                            id: 0,
+                            retries: 0,
+                            nickname: nickname.clone(),
+                            task,
+                        }).clone(),
+                    )),
+
+                    Entry::Occupied(mut entry) => {
+                        entry.insert(Job {
+                            id: entry.get().id + 1,
+                            retries: 0,
+                            nickname: nickname.clone(),
+                            task,
+                        });
+                    }
+                }
+            }
+
+            else => (),
+        }
+    }
+}
+
+/// An discriminant for message queues: messages with the same queue ID will be delivered
+/// in order.
+#[derive(Copy, Clone, PartialEq, Eq, Hash)]
+struct QueueId {
+    sender: ChainId,
+    recipient: ChainId,
+    is_update: bool,
+}
+
+impl QueueId {
+    /// Returns a discriminant for the message's queue.
+    fn new(request: &CrossChainRequest) -> Self {
```

Comment:
nit: Personally I think `from` would be more appropriate.

Reply:
For me I think I usually expect `from` to be an injection.

Reply:
Here it's literally mapping from `CrossChainRequest` to a `QueueId`.

Reply:
Right, which is definitely not an injection :)

Reply:
That's why I didn't suggest implementing `From`/`Into` but rename `new` into `from` which makes it clear that `QueueId` is constructed _from_ a `CrossChainRequest`.

Reply:
Not gonna insist though :)